contracts/modules/token/TokenController.sol

```diff
@@ -310,13 +311,68 @@ contract TokenController is ITokenController, LockHandler, MasterAwareV2 {
     }
   }
 
-  /**
-  * @dev Returns tokens locked for a specified address for a
-  *    specified reason
-  *
-  * @param _of The address whose tokens are locked
-  * @param _reason The reason to query the lock tokens for
-  */
+  /// @notice Withdraws NXM from the Nexus platform based on specified options.
+  /// @dev    Ensure the NXM is available and not locked before withdrawal. Only set flags in `WithdrawNxmOptions` for
+  ///         withdrawable NXM. Reverts if some of the NXM being withdrawn is locked or unavailable.
+  /// @param stakingPoolDeposit  Details for withdrawing from staking pools.
+  /// @param coverNotes          Details for withdrawing V1 cover notes.
+  /// @param batchSize           The maximum number of iterations to avoid unbounded loops when withdrawing
+  ///                            governance and/or assessment rewards.
+  /// @param withdrawOptions     Options specifying which NXM types to withdraw, set flags to true to include specific
+  ///                            withdrawable NXM.
+  function withdrawNXM(
+    StakingPoolDeposit calldata stakingPoolDeposit,
+    V1CoverNotes calldata coverNotes,
+    uint batchSize,
+    WithdrawNxmOptions calldata withdrawOptions
+  ) external whenNotPaused {
+
+    // assessment stake
+    if (withdrawOptions.assessmentStake) {
+      IAssessment _assessment = assessment();
+      (uint96 amount, , ) = _assessment.stakeOf(msg.sender);
+      _assessment.unstakeFor(msg.sender, amount, msg.sender);
+    }
+
+    // assessment and governance rewards
+    if (withdrawOptions.governanceRewards || withdrawOptions.assessmentRewards) {
+      withdrawPendingRewards(
+        msg.sender,
+        withdrawOptions.governanceRewards,
+        withdrawOptions.assessmentRewards,
+        batchSize
+      );
+    }
+
+    // staking pools rewards and stake (including manager rewards)
+    if (withdrawOptions.stakingPoolStake || withdrawOptions.stakingPoolRewards) {
```

Comment:
I'm considering to remove these flags for simplicity. The cases are as follows:
1. No stake/rewards - just pass an empty array.
2. Expired stake:
   a. withdrawing both is achieved by default
   b. withdrawing just the stake doesn't necessarily make sense - no reason to leave the rewards or withdraw them separately
   c. withdrawing just the rewards would (kinda) make sense if the user wants to withdraw the rewards but extend their stake, but extend calls withdraw+deposit so technically it would have the same effect as withdrawing and then depositing (just not atomic), so no point to do this either
3. Non-expired stake: only rewards are withdrawable, to avoid withdrawing them just pass an empty array (or skip the id)

In conclusion all flows seem to be covered and we can drop the flags

Reply:
ah and should pass `true, true` to the staking pool. if the stake is not expired there's a check there and it will be skipped anyway - no reverts there.

Reply:
ah, actually there is a slight edge case - withdrawing expired stake if the manager has voted in governance in which case the staker cannot withdraw the stake (but they can withdraw the rewards). I'd say it makes sense to skip both and withdraw both later, i.e. consider rewards unwithdrawable just like the stake.

Reply:
to withdraw stake / normal rewards we need:
* tokenIds
* corresponding trancheIds

```solidity
uint poolId = stakingNFT.stakingPoolOf(tokenId);
stakingPool(poolId).withdraw(tokenId, true, true, trancheIds);
```

to withdraw manager rewards we need:
* managed poolIds (since `stakingNFT.stakingPoolOf(0)` will not work)
* corresponding trancheIds for token 0

```solidity
stakingPool(poolId).withdraw(0, true, true, trancheIds);
```

Reply:
so I ended up with this struct to allow user to withdraw both normal / manager rewards at the same time:
```solidity
  struct StakingPoolDeposit {
    uint[] tokenIds;
    uint[][] tokenTrancheIds;
    uint[] managedPoolIds;
    uint[][] managerRewardsTrancheIds;
  }
```

Reply:
so I ended up with this struct:
```solidity
struct StakingPoolDeposit {
    uint[] tokenIds;
    uint[][] tokenTrancheIds;
    uint[] managedPoolIds;
    unit[][] managarRewardsTrancheIds
}
```
its possible to drop `managarRewardsTrancheIds` and re-use the `tokenTrancheIds` but it gets a bit complicated / un-intuitive for the caller:
```solidity
{
  tokenIds: [1,2,3,0,0,5] // if tokenId != 0 fetch poolId, else grab from managedPoolIds
  managedPoolIds: [0,0,0,1,2,0] // <-- only tokenId 0s have corresponding valid poolIds (index 3, 4)
  trancheIds: [[221],[222],[223],[224],[225]]
}
```

Reply:
as discussed it will be implemented as 2 separate structs:

```solidity
  struct StakingPoolDeposit {
    uint tokenId;
    uint[] trancheIds;
  }

  struct StakingPoolManagerReward {
    uint poolId;
    uint[] trancheIds;
  }
```